docutils cli: fail-fast validation of paths

boneskull opened the pull request:
This adds up-front checking of user-provided custom paths to avoid doing a bunch of work and then
failing later ("fail fast").

For example, a user may provide a custom path to `mkdocs.yml` on the CLI:

```bash
appium-docs build --mkdocs-yml /some/bad/path/to/mkdocs.yml
```

In the above case, `appium-docs` will check if `/some/bad/path/to/mkdocs.yml` exists _before_ running
TypeDoc and MkDocs.

An alternative solution would be to just ignore it and use the default path.  This was _clearly_
not the user's intent (because they provided a custom path), so it's a poor solution.

boneskull committed:
fix(appium,support): fs.readPackageJsonFrom() returns proper type

This function always returns a `NormalizedPackageJson`, so let's be explicit about that.

boneskull committed:
feat(docutils): better path validation

This adds up-front checking of user-provided custom paths to avoid doing a bunch of work and then failing later ("fail fast")

boneskull commented:
Current dependencies on/for this PR:
* master
  * **PR #18261** <a href="https://app.graphite.dev/github/pr/appium/appium/18261" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #18265** <a href="https://app.graphite.dev/github/pr/appium/appium/18265" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #18266** <a href="https://app.graphite.dev/github/pr/appium/appium/18266" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #18280** <a href="https://app.graphite.dev/github/pr/appium/appium/18280" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/18280?utm_source=stack-comment).

mykola-mokhnach commented on packages/docutils/lib/cli/check.ts:
```diff
@@ -0,0 +1,88 @@
+ */
+async function filterMissing(paths: MissingFileData[]): Promise<MissingFileData[]> {
+  const exists = await Promise.all(
+    paths.map(async ({id, path}) => ({id, path, exists: await fs.exists(path)}))
```
I did not know that .map accepts async functions

jlipps replied:
there's nothing magical going on here. async functions are just functions that return a promise when called. so this use of `.map` will produce a list of promises, which are resolved by `await Promise.all`.

mykola-mokhnach commented on packages/docutils/lib/cli/check.ts:
```diff
@@ -0,0 +1,88 @@
+    paths.map(async ({id, path}) => ({id, path, exists: await fs.exists(path)}))
+  );
+  const results = _.reject(exists, 'exists');
+  return _.map(results, (result) => _.omit(result, 'exists'));
```
I assume bluebird's filter could do this all in one line

boneskull replied:
I could do this on one line with lodash too, but it was too clever

mykola-mokhnach commented on packages/docutils/lib/cli/check.ts:
```diff
@@ -0,0 +1,88 @@
+ * @param opts Options object for a yargs command
+ * @param group Group name to filter on
+ * @param argv User-provided args
+ * @returns `true` if all paths exist or otherwise an error message
```
would it make more sence to just throw an error instead of returning a string?

boneskull replied:
we could do that, but I figure that'd produce a stack trace somewhere, and it's unneeded for checking parameters.  the return value of these functions is just what yargs expects.

mykola-mokhnach commented on packages/docutils/lib/cli/check.ts:
```diff
@@ -0,0 +1,88 @@
+
+  log.debug(
+    'Checking for existence of %s: %s',
+    pluralize('path', pathsToCheck.length),
```
there is util.pluralize function in the support module

boneskull replied:
good catch -- I started depending directly on `pluralize` in `@appium/typedoc-plugin-appium` because I didn't need to pull in all of `@appium/support` (though it's ostensibly already installed).  guess I forgot why I was doing what I was doing

mykola-mokhnach commented on packages/docutils/lib/cli/command/init.ts:
```diff
@@ -144,17 +157,21 @@ const opts = Object.freeze({
+  builder(yargs) {
+    return yargs
+      .options(opts)
+      .check(async (argv) => checkMissingPaths(opts, InitCommandGroup.Paths, argv));
```
do we need await in the lambda body?

boneskull replied:
we don't have lint support for typescript installed yet (see #18159) so idk.  I was never a fan of that rule

mykola-mokhnach commented on packages/docutils/lib/cli/command/validate.ts:
```diff
@@ -76,25 +84,25 @@ const opts = Object.freeze({
+      if (!argv.python && !argv.typedoc && !argv.typescript && !argv.mkdocs) {
+        return 'No validation targets specified; one or more of --python, --typescript, --typedoc or --mkdocs must be provided';
+      }
+      return checkMissingPaths(opts, ValidateCommandGroup.Paths, argv);
```
await?

boneskull replied:
more about this: my rationale for not using it is that it does not matter what `checkMissingPaths` returns because the function consuming it is `async`; adding `await` here without a `try/catch` serves no practical purpose other than causing churn if it did happen to become a sync function (because keeping a useless `await` will be slow in a hot path and could be the source of subtle bugs).  

anyway, I'm going to leave this for now and I can make my case when we add `typescript-eslint` support.

boneskull committed:
chore(docutils): review updates
- Use `util.pluralize()` from `@appium/support` instead of `pluralize` directly
- Write some more docstrings
